File: tests/benchmarks/node/bench.js
const processArgs = (() => {
  const out = { scale: 500 };
  for (let i = 2; i < process.argv.length; i++) {
    const a = process.argv[i];
    if (a === "--scale") out.scale = parseInt(process.argv[++i], 10);
  }
  return out;
})();

function nowMs() {
  return process.hrtime.bigint();
}

function benchLoop(n) {
  const t0 = nowMs();
  let s = 0;
  for (let i = 0; i < n; i++) s += 1;
  const t1 = nowMs();
  return { case: "loop", scale: n, result: s, duration_ms: Number(t1 - t0) / 1e6 };
}

function benchDict(n) {
  const t0 = nowMs();
  const d = {};
  for (let i = 0; i < n; i++) d["k" + i] = i;
  let s = 0;
  for (let i = 0; i < n; i++) s += d["k" + i];
  const t1 = nowMs();
  return { case: "dict", scale: n, result: s, duration_ms: Number(t1 - t0) / 1e6 };
}

function benchDictIntKey(n) {
  const t0 = nowMs();
  const d = {};
  for (let i = 0; i < n; i++) d[i] = i;
  let s = 0;
  for (let i = 0; i < n; i++) s += d[i];
  const t1 = nowMs();
  return { case: "dict-intkey", scale: n, result: s, duration_ms: Number(t1 - t0) / 1e6 };
}

function benchString(n) {
  const t0 = nowMs();
  const arr = new Array(n).fill(0).map((_, i) => String(i));
  const s = arr.join(",");
  const parts = s.split(",");
  const t1 = nowMs();
  return { case: "string", scale: n, result: parts.length, duration_ms: Number(t1 - t0) / 1e6 };
}

function benchStringBuilder(n) {
  const t0 = nowMs();
  const parts = new Array(n);
  for (let i = 0; i < n; i++) parts[i] = String(i);
  const s = parts.join(",");
  const t1 = nowMs();
  return { case: "string-builder", scale: n, result: s.length, duration_ms: Number(t1 - t0) / 1e6 };
}

function benchDictHot(n) {
  const d = {};
  for (let i = 0; i < n; i++) d["k" + i] = i;
  const hot = "k" + (n >> 1);
  const t0 = nowMs();
  let s = 0;
  for (let i = 0; i < n; i++) s += d[hot];
  const t1 = nowMs();
  return { case: "dict-hot", scale: n, result: s, duration_ms: Number(t1 - t0) / 1e6 };
}

class Obj {
  constructor() { this.base = 1; }
  method(x) { return x + this.base; }
}

function benchStructMethod(n) {
  const o = new Obj();
  const t0 = nowMs();
  let s = 0;
  for (let i = 0; i < n; i++) s += o.method(i);
  const t1 = nowMs();
  return { case: "struct-method", scale: n, result: s, duration_ms: Number(t1 - t0) / 1e6 };
}

function benchFuncCall(n) {
  function f(x) { return x + 1; }
  const t0 = nowMs();
  let s = 0;
  for (let i = 0; i < n; i++) s += f(i);
  const t1 = nowMs();
  return { case: "func-call", scale: n, result: s, duration_ms: Number(t1 - t0) / 1e6 };
}

function benchBranchHeavy(n) {
  const t0 = nowMs();
  let s = 0;
  for (let i = 0; i < n; i++) {
    if ((i & 1) === 0) s += 1;
    else s -= 1;
  }
  const t1 = nowMs();
  return { case: "branch-heavy", scale: n, result: s, duration_ms: Number(t1 - t0) / 1e6 };
}

function benchListPushPop(n) {
  const t0 = nowMs();
  const l = [];
  for (let i = 0; i < n; i++) l.push(i);
  for (let i = 0; i < n; i++) l.pop();
  const t1 = nowMs();
  return { case: "list-push-pop", scale: n, result: l.length, duration_ms: Number(t1 - t0) / 1e6 };
}

function benchDictMiss(n) {
  const d = {};
  for (let i = 0; i < n; i++) d["k" + i] = i;
  const t0 = nowMs();
  let s = 0;
  for (let i = 0; i < n; i++) s += (d["x" + i] || 0);
  const t1 = nowMs();
  return { case: "dict-miss", scale: n, result: s, duration_ms: Number(t1 - t0) / 1e6 };
}

function benchDictUpdateHot(n) {
  const d = { hot: 0 };
  const t0 = nowMs();
  for (let i = 0; i < n; i++) d.hot = i;
  const t1 = nowMs();
  return { case: "dict-update-hot", scale: n, result: d.hot, duration_ms: Number(t1 - t0) / 1e6 };
}

function benchStringUnicode(n) {
  const t0 = nowMs();
  const parts = ["こんにちは", "世界🌏", "Xu", "脚本"];
  const arr = new Array(n);
  for (let i = 0; i < n; i++) arr[i] = parts[i % parts.length];
  const s = arr.join(",");
  const out = s.split(",");
  const t1 = nowMs();
  return { case: "string-unicode", scale: n, result: out.length, duration_ms: Number(t1 - t0) / 1e6 };
}

function benchStringScan(n) {
  const s = new Array(n).fill(0).map((_, i) => String(i)).join(",");
  const t0 = nowMs();
  const c1 = s.includes("999");
  const c2 = s.startsWith("0,1");
  const c3 = s.endsWith(String(n - 1));
  const t1 = nowMs();
  return { case: "string-scan", scale: n, result: (c1 && c2 && c3) ? 1 : 0, duration_ms: Number(t1 - t0) / 1e6 };
}

function runCase(fn, n, warms, repeat) {
  for (let i = 0; i < warms; i++) fn(n);
  let best = null;
  let last = null;
  for (let i = 0; i < repeat; i++) {
    const item = fn(n);
    last = item;
    const d = item.duration_ms;
    if (best === null || d < best) best = d;
  }
  last.duration_ms = best;
  last.rss_bytes = process.memoryUsage().rss;
  return last;
}

const n = processArgs.scale;
const warms = parseInt(process.env.WARMS || "0", 10);
const repeat = parseInt(process.env.REPEAT || "1", 10);
const fns = [
  benchLoop,
  benchDict,
  benchDictIntKey,
  benchString,
  benchStringBuilder,
  benchDictHot,
  benchStructMethod,
  benchFuncCall,
  benchBranchHeavy,
  benchListPushPop,
  benchDictMiss,
  benchDictUpdateHot,
  benchStringUnicode,
  benchStringScan,
];
for (const fn of fns) {
  const item = runCase(fn, n, warms, repeat);
  console.log(JSON.stringify(item));
}

File: tests/benchmarks/python/bench.py
import argparse
import time
import json
import os
import resource

def bench_loop(n):
    t0 = time.perf_counter()
    s = 0
    for i in range(n):
        s += 1
    t1 = time.perf_counter()
    return {"case": "loop", "scale": n, "result": s, "duration_ms": (t1 - t0) * 1000.0}

def bench_dict(n):
    t0 = time.perf_counter()
    d = {}
    for i in range(n):
        d[f"k{i}"] = i
    s = 0
    for i in range(n):
        s += d[f"k{i}"]
    t1 = time.perf_counter()
    return {"case": "dict", "scale": n, "result": s, "duration_ms": (t1 - t0) * 1000.0}

def bench_string(n):
    t0 = time.perf_counter()
    arr = [str(i) for i in range(n)]
    s = ",".join(arr)
    parts = s.split(",")
    t1 = time.perf_counter()
    return {"case": "string", "scale": n, "result": len(parts), "duration_ms": (t1 - t0) * 1000.0}

def bench_dict_intkey(n):
    t0 = time.perf_counter()
    d = {}
    for i in range(n):
        d[i] = i
    s = 0
    for i in range(n):
        s += d[i]
    t1 = time.perf_counter()
    return {"case": "dict-intkey", "scale": n, "result": s, "duration_ms": (t1 - t0) * 1000.0}

def bench_string_builder(n):
    t0 = time.perf_counter()
    parts = []
    for i in range(n):
        parts.append(str(i))
    s = ",".join(parts)
    t1 = time.perf_counter()
    return {"case": "string-builder", "scale": n, "result": len(s), "duration_ms": (t1 - t0) * 1000.0}

def bench_dict_hot(n):
    d = {f"k{i}": i for i in range(n)}
    hot = f"k{n//2}"
    t0 = time.perf_counter()
    s = 0
    for _ in range(n):
        s += d[hot]
    t1 = time.perf_counter()
    return {"case": "dict-hot", "scale": n, "result": s, "duration_ms": (t1 - t0) * 1000.0}

class Obj:
    def __init__(self):
        self.base = 1
    def method(self, x):
        return x + self.base

def bench_struct_method(n):
    o = Obj()
    t0 = time.perf_counter()
    s = 0
    for i in range(n):
        s += o.method(i)
    t1 = time.perf_counter()
    return {"case": "struct-method", "scale": n, "result": s, "duration_ms": (t1 - t0) * 1000.0}

def bench_func_call(n):
    def f(x): return x + 1
    t0 = time.perf_counter()
    s = 0
    for i in range(n):
        s += f(i)
    t1 = time.perf_counter()
    return {"case": "func-call", "scale": n, "result": s, "duration_ms": (t1 - t0) * 1000.0}

def bench_branch_heavy(n):
    t0 = time.perf_counter()
    s = 0
    for i in range(n):
        if (i & 1) == 0:
            s += 1
        else:
            s -= 1
    t1 = time.perf_counter()
    return {"case": "branch-heavy", "scale": n, "result": s, "duration_ms": (t1 - t0) * 1000.0}

def bench_list_push_pop(n):
    t0 = time.perf_counter()
    l = []
    for i in range(n):
        l.append(i)
    for i in range(n):
        l.pop()
    t1 = time.perf_counter()
    return {"case": "list-push-pop", "scale": n, "result": len(l), "duration_ms": (t1 - t0) * 1000.0}

def bench_dict_miss(n):
    d = {f"k{i}": i for i in range(n)}
    t0 = time.perf_counter()
    s = 0
    for i in range(n):
        s += d.get(f"x{i}", 0)
    t1 = time.perf_counter()
    return {"case": "dict-miss", "scale": n, "result": s, "duration_ms": (t1 - t0) * 1000.0}

def bench_dict_update_hot(n):
    d = {"hot": 0}
    t0 = time.perf_counter()
    for i in range(n):
        d["hot"] = i
    t1 = time.perf_counter()
    return {"case": "dict-update-hot", "scale": n, "result": d["hot"], "duration_ms": (t1 - t0) * 1000.0}

def bench_string_unicode(n):
    t0 = time.perf_counter()
    parts = ["こんにちは", "世界🌏", "Xu", "脚本"] * (n // 4)
    s = ",".join(parts)
    out = s.split(",")
    t1 = time.perf_counter()
    return {"case": "string-unicode", "scale": n, "result": len(out), "duration_ms": (t1 - t0) * 1000.0}

def bench_string_scan(n):
    s = ",".join([str(i) for i in range(n)])
    t0 = time.perf_counter()
    c1 = "999" in s
    c2 = s.startswith("0,1")
    c3 = s.endswith(str(n - 1))
    t1 = time.perf_counter()
    return {"case": "string-scan", "scale": n, "result": int(c1 and c2 and c3), "duration_ms": (t1 - t0) * 1000.0}

def bench_closure_create(n):
    t0 = time.perf_counter()
    total = 0
    for i in range(n):
        captured = i
        f = lambda x, c=captured: x + c
        total += f(1)
    t1 = time.perf_counter()
    return {"case": "closure-create", "scale": n, "result": total, "duration_ms": (t1 - t0) * 1000.0}

def bench_closure_call(n):
    captured = 42
    f = lambda x: x + captured
    t0 = time.perf_counter()
    total = 0
    for i in range(n):
        total += f(i)
    t1 = time.perf_counter()
    return {"case": "closure-call", "scale": n, "result": total, "duration_ms": (t1 - t0) * 1000.0}

def run_case(fn, n, warms, repeat):
    for _ in range(warms):
        fn(n)
    total = 0.0
    last = None
    for _ in range(repeat):
        item = fn(n)
        last = item
        total += item["duration_ms"]
    rss = resource.getrusage(resource.RUSAGE_SELF).ru_maxrss
    last["duration_ms"] = total / repeat
    last["rss_bytes"] = int(rss) if isinstance(rss, (int, float)) else 0
    return last

def main():
    ap = argparse.ArgumentParser()
    ap.add_argument("--scale", type=int, default=500)
    args = ap.parse_args()
    warms = int(os.environ.get("WARMS", "0") or "0")
    repeat = int(os.environ.get("REPEAT", "1") or "1")
    fns = [
        bench_loop,
        bench_dict,
        bench_dict_intkey,
        bench_string,
        bench_string_builder,
        bench_dict_hot,
        bench_struct_method,
        bench_func_call,
        bench_branch_heavy,
        bench_list_push_pop,
        bench_dict_miss,
        bench_dict_update_hot,
        bench_string_unicode,
        bench_string_scan,
        bench_closure_create,
        bench_closure_call,
    ]
    for fn in fns:
        item = run_case(fn, args.scale, warms, repeat)
        print(json.dumps(item, ensure_ascii=False))

if __name__ == "__main__":
    main()
